src/Offer/Popularity/InMemoryPopularityRepository.php

```diff
@@ -0,0 +1,23 @@
+<?php
+
+declare(strict_types=1);
+
+namespace CultuurNet\UDB3\Offer\Popularity;
+
+final class InMemoryPopularityRepository implements PopularityRepository
+{
```

Comment:
Just out of curiosity, what's the reason of creating a concrete class for testing purposes instead of mocking?

Reply:
I like the clarity compared to using mocks though. If it's only used in tests, I would suggest moving it closer to the tests, potentially even as an [anonymous class](https://www.php.net/manual/en/language.oop5.anonymous.php) that implements the interface.

Reply:
Not sure if this clarity out-weights the extra code to maintain. I'm personally still in favour of mocking, but that can change of course 😉

Maybe it's also dependent on the size if the interface. This is a small interface so the concrete class is also very small. But with bigger interfaces that can result in more code. Although a big interface is maybe a code smell.

Reply:
Maybe the `InMemory` implementation of the `DocumentRepository` (used in the same class) is a better example. It's a lot less code to maintain compared to having mocks in all the tests it's being used. But that's also a widely-used interface. (We already had that `InMemory` implementation long before btw.)

I personally also think it's clearer to have very simple implementations compared to mocks, and in the case of repositories it's usually simple to implement with an array.

Maybe also less chance of mocking something incorrectly? But that can be up for debate since you could also implement the InMemory implementation incorrectly.

But in this case, it also just felt like less work to implement it once compared to writing mock expectations in every test.

For more complex interfaces mocks are definitely the way to go of course.

And in the end, it's probably subjective as well. (And trivial)

I like the suggestion of moving it closer to the tests and/or using anonymous classes @jdrieghe . I was actually in doubt whether to place the call inside the `tests` or `src` folder, but ended up going with `src` because we also have other `InMemory` implementations for tests there. But that was maybe to work around a limitation when they were still in `udb3-php` and if they would be in the `tests` folder there, we couldn't load them inside the tests in `udb3-silex` because composer wouldn't autoload the classes in the `tests` folders of dependencies. So yeah we could move them to `tests` now. But maybe that's pretty minor and not necessarily worth the effort right now?

Reply:
> Not sure if this clarity out-weights the extra code to maintain.

It's 23 lines of code and 1 line of code per test (calling the `saveScore` method), compared to ~4 lines of code per test to mock the methods:

```
$this->popularityRepository->expects($this->once())
    ->method('get')
    ->with($offerId, $includeMetadata)
    ->willReturn(...)
```

So in this case of only 3 tests, it's a little bit more code yes (`23+3*1 = 26` vs `3*4 = 12`, so 14 lines of code extra), but negligible I'd say. And in cases where there'd be more tests (like all the tests where we use `InMemoryDocumentRepository`) it would actually save a lot of code.

Plus if the interface ever changes it's easier to refactor the implementation than to refactor multiple mocks imho.

Reply:
I'm convinced now! 😀

> It's a lot less code to maintain compared to having mocks in all the tests where it's being used
I was too focused on this pull request. But if we make more classes dependent on `PopularityRepository` that would indeed require lots of duplicated mock code which can be avoided with this concrete implementation.